src/main/controller/PromotionController.ts

```diff
@@ -204,4 +196,70 @@ export class PromotionController {
       return next(e);
     }
   };
+
+  /**
+   * Get the restaurant details for a promotion
+   */
+  getRestaurantDetails = async (
+    request: Request,
+    response: Response,
+    next: NextFunction
+  ): Promise<any> => {
+    try {
+      await getManager().transaction(async (transactionalEntityManager) => {
+        let result: Place = {};
+        const id = await IdValidation.schema.validateAsync(request.params.id, {
+          abortEarly: false,
+        });
+
+        const promotion = await transactionalEntityManager
+          .getCustomRepository(PromotionRepository)
+          .findOneOrFail(id);
+
+        // placeId may be empty string if this promotion has previously resulted in a NOT_FOUND even after refresh request
+        if (promotion.placeId) {
+          const placeDetailsResponseData = await this.googlePlaceService.getRestaurantDetails(
+            promotion.placeId
+          );
+          result = placeDetailsResponseData.result ?? {};
+
+          if (placeDetailsResponseData.status === Status.NOT_FOUND) {
+            result = await this.handlePlaceIdNotFound(
+              promotion.placeId,
+              id,
+              transactionalEntityManager
+            );
+          }
+        }
+
+        return response.status(200).send(result);
+      });
+    } catch (e) {
+      return next(e);
+    }
+  };
+
+  /**
+   * Handle NOT_FOUND case for getting restaurant details of a placeId for a promotion
+   * 1. Issue refresh request for the placeId
+   * 2. Store new placeId in DB, even if placeId is empty string
+   * @param placeId the placeId of the promotion
+   * @param id the id of the promotion
+   * @param transactionalEntityManager
+   * @return Place - the restaurant details which may be empty
+   * */
+  private async handlePlaceIdNotFound(
+    placeId: string,
+    id: string,
+    transactionalEntityManager: EntityManager
+  ): Promise<Place> {
+    const refreshResult = await this.googlePlaceService.refreshPlaceId(placeId);
+
+    // update DB with new placeId, even if placeId is empty string
+    await transactionalEntityManager
+      .getCustomRepository(PromotionRepository)
+      .update({ id }, { placeId: refreshResult.placeId });
+
+    return refreshResult.restaurantDetails;
```

Comment:
Should we also set the placeIDs of all the other promotions associated with the same restaurant to empty as well? And also each of their associated restaurant entity?

Reply:
That's a very good point, yes. I think once I have refactored it into a oneToMany relationship this issue won't exist, but that's a very good point